src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzer.cs

```diff
@@ -0,0 +1,169 @@
+
+        internal static DiagnosticDescriptor CastRule = DiagnosticDescriptorHelper.Create(RuleId,
+                                                                             s_localizableCastTitle,
+                                                                             s_localizableCastMessage,
+                                                                             DiagnosticCategory.Reliability,
+                                                                             RuleLevel.IdeSuggestion,
+                                                                             s_localizableCastDescription,
+                                                                             isPortedFxCopRule: false,
+                                                                             isDataflowRule: false);
+
+        internal static DiagnosticDescriptor OfTypeRule = DiagnosticDescriptorHelper.Create(RuleId,
+                                                                             s_localizableOfTypeTitle,
+                                                                             s_localizableOfTypeMessage,
+                                                                             DiagnosticCategory.Reliability,
+                                                                             RuleLevel.IdeSuggestion,
+                                                                             description: s_localizableOfTypeDescription,
+                                                                             isPortedFxCopRule: false,
+                                                                             isDataflowRule: false);
+
+        private static readonly ImmutableArray<(string MethodName, DiagnosticDescriptor Rule)> methodMetadataNames = ImmutableArray.Create(
+            ("OfType", OfTypeRule),
+            ("Cast", CastRule)
+        );
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+            => ImmutableArray.Create(OfTypeRule, CastRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemLinqEnumerable, out var enumerableType))
+                {
+                    return;
+                }
+
+                var methodRuleDictionary = methodMetadataNames
+                    .SelectMany(m => enumerableType.GetMembers(m.MethodName)
+                                                   .OfType<IMethodSymbol>()
+                                                   .Where(method => method.TypeParameters.HasExactly(1)
+                                                                 && method.Parameters.HasExactly(1))
```

Comment:
@mavasani any thoughts on the best way of locating a particular member?

Reply:
Am I understanding that would cause this not to trigger without the extension method syntax?

ie. is there a reason why there should be a difference between
 `(new string[] { "not an int" }).Cast<int>();` and
 `Enumerable.Cast<int>(new string[] { "not an int" });`?

Reply:
> Am I understanding that would cause this not to trigger without the extension method syntax?

Yes, and it is better to only restrict to the existing implementation of `Enumerable.Cast<T>/OfType<T>` which takes `this` parameter. There is no other overload for Enumerable.Cast<T>,  and if an overload is added in the future we don't know that overload would have the same behavior and needs to warn

Reply:
> ie. is there a reason why there should be a difference between

I don't see any difference between them for the analyzer

Reply:
ahh, I was confusing method definition and call site. I used `method.IsExtensionMethod` instead.